### PR TITLE
Add S3 security mapping for native S3

### PR DESCRIPTION
**Pull request by electrum**
## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Object storage
* Add S3 security mapping for native S3. ({issue}`22559`)
```

### CODE REVIEW

**Comment by electrum** [comment]
@mosabua This is compatible with the legacy S3 feature and uses the same config properties (minus the `hive.` prefix). How do you want to handle the docs? Should we just copy/paste?

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemFactory.java [hunk 1]
```diff
@@ -15,160 +15,42 @@
...
+    private final S3FileSystemLoader loader;
     private final S3Client client;
     private final S3Context context;
-    private final ExecutorService uploadExecutor;
+    private final Executor uploadExecutor;
 
     @Inject
     public S3FileSystemFactory(OpenTelemetry openTelemetry, S3FileSystemConfig config)
     {
-        S3ClientBuilder s3 = S3Client.builder();
```
Comment (nineinchnick) [inline lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemFactory.java:60]: Would be easier to review if moving of this code to the loader was extracted to a separate commit.

**Review by nineinchnick — APPROVED** [review]
Are the tests based on some existing ones in the legacy FS? I haven't reviewed them for coverage.

**Comment by mosabua** [comment]
> @mosabua This is compatible with the legacy S3 feature and uses the same config properties (minus the `hive.` prefix). How do you want to handle the docs? Should we just copy/paste?

Yeah .. we can just have it in the new docs as copy.. then cleaning up later will be easier. You want to do that or should I send a separate PR @electrum ?

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SecurityMappingsFileSource.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+class S3SecurityMappingsFileSource
+        implements Supplier<S3SecurityMappings>
+{
+    private final File configFile;
+    private final String jsonPointer;
+
+    @Inject
+    public S3SecurityMappingsFileSource(S3SecurityMappingConfig config)
+    {
+        this.configFile = config.getConfigFilePath().map(File::new)
```
[the remark this answers is not in the document]
Comment (electrum) [reply lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SecurityMappingsFileSource.java:33]: It's not needed because the code directly returns a non-null `Optional`. The validation annotations are for the Bean Validation framework to validate user-supplied values.

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SecurityMappingsFileSource.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                .orElseThrow(() -> new IllegalArgumentException("configFilePath not set"));
+        if (!configFile.exists()) {
```
[the remark this answers is not in the document]
Comment (electrum) [reply lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SecurityMappingsFileSource.java:35]: We can't use `@FileExists` because the file path can also be a URL. But we could use `@AssertTrue`. I'll add that.

[the remark this answers is not in the document]
Comment (electrum) [reply lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SecurityMappingsFileSource.java:35]: Actually, I'm going to make two configs.

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemFactory.java (hunk 1 above)
Comment (electrum) [reply lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemFactory.java:60]: I completely agree, and tried to do that originally, but the code is restructured significantly. I didn't see a good way to do a clean refactoring.

**Comment by electrum** [comment]
I copied over the documentation and cleaned it up a bit.

**Review by anusudarsan — APPROVED** [review]

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 4]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
 * - `s3.external-id`
   - External ID for the IAM role trust policy when connecting to S3.
 :::
+
+## Security mapping
+
+Trino supports flexible security mapping for S3, allowing for separate
+credentials or IAM roles for specific users or buckets/paths. The IAM role
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:130]: can we somehow avoid  "bucket/paths". Maybe even by using "users or buckets or paths" .. even though they are equivalent. What do we do in the rest of the docs or what is happening in the S3 docs for this?

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 5]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+for a specific query can be selected from a list of allowed roles by providing
+it as an *extra credential*.
+
+Each security mapping entry may specify one or more match criteria. If multiple
+criteria are specified, all criteria must match. Available match criteria:
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:135]:
```suggestion
criteria are specified, all criteria must match. The following match criteria are available:
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 6]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+
+- `user`: Regular expression to match against username. Example: `alice|bob`
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:137]: do we need to say what type of regex?

Comment (mosabua) [reply docs/src/main/sphinx/object-storage/file-system-s3.md:137]: And we should explain that the examples uses an OR and therefore matches for alice and bob

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 7]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+- `group`: Regular expression to match against any of the groups that the user
+  belongs to. Example: `finance|sales`
+- `prefix`: S3 URL prefix. It can specify an entire bucket or a path within a
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:140]:
```suggestion
- `prefix`: S3 URL prefix. You can specify an entire bucket or a path within a
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 8]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+  bucket. The URL must start with `s3://` but will also match `s3a` or `s3n`.
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:141]:
```suggestion
  bucket. The URL must start with `s3://` but also matches for `s3a` or `s3n`.
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 9]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+  Example: `s3://bucket-name/abc/xyz/`
+
+The security mapping must provide one or more configuration settings:
+
+- `accessKey` and `secretKey`: AWS access key and secret key. This overrides
+  any globally configured credentials, such as access key or instance credentials.
+- `iamRole`: IAM role to use if no user provided role is specified as an
+  extra credential. This overrides any globally configured IAM role. This role
+  is allowed to be specified as an extra credential, although specifying it
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:150]:
```suggestion
  can be specified as an extra credential, although specifying it
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 10]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+  explicitly has no effect, as it would be used anyway.
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:151]:
```suggestion
  explicitly has no effect, as it is used anyway.
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 11]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+- `roleSessionName`: Optional role session name to use with `iamRole`. This can only
+  be used when `iamRole` is specified. If `roleSessionName` includes the string
+  `${USER}`, then the `${USER}` portion of the string will be replaced with the
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:154]:
```suggestion
  `${USER}`, then the `${USER}` portion of the string is replaced with the
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 12]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+- `allowedIamRoles`: IAM roles that are allowed to be specified as an extra
+  credential. This is useful because a particular AWS account may have permissions
+  to use many roles, but a specific user should only be allowed to use a subset
+  of those roles.
+- `kmsKeyId`: ID of KMS-managed key to be used for client-side encryption.
+- `allowedKmsKeyIds`: KMS-managed key IDs that are allowed to be specified as an extra
+  credential. If list cotains `*`, then any key can be specified via extra credential.
+- `endpoint`: The S3 storage endpoint server. This optional property can be used
+  to override S3 endpoints on a per-bucket basis.
+- `region`: The region S3 client should connect to. This optional property can be used
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:166]:
```suggestion
- `region`: The S3 region to connect to. This optional property can be used
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 13]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+  to override S3 regions on a per-bucket basis.
+
+The security mapping entries are processed in the order listed in the configuration
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:169]: maybe "JSON configuration".

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 14]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+JSON. More specific mappings should thus be specified before less specific mappings.
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:170]:
```suggestion
JSON. Therefore, more specific mappings must be specified before less specific mappings.
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 15]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+For example, the mapping list might have URL prefix `s3://abc/xyz/` followed by
+`s3://abc/` to allow different configuration for a specific path within a bucket
+than for other paths within the bucket. You can specify the default configuration
+by not including any match criteria for the last entry in the list.
+
+In addition to the rules above, the default mapping can contain the optional
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:176]:
```suggestion
In addition to the preceding rules, the default mapping can contain the optional
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 16]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+`useClusterDefault` boolean property set to true to use the default S3 configuration.
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:177]:
```suggestion
`useClusterDefault` boolean property set to `true` to use the default S3 configuration.
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 17]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+It cannot be used with any other configuration settings.
+
+If no mapping entry matches and no default is configured, access is denied.
+
+The configuration JSON can either be retrieved from a file via
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:182]:
```suggestion
The configuration JSON can be retrieved from a file via
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 18]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+    },
+    {
+      "iamRole": "arn:aws:iam::123456789101:role/default"
+    }
+  ]
+}
+```
+
+:::{list-table}
+:header-rows: 1
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:235]: Maybe add percentages for the columns depending on how it renders now.

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 19]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+      "iamRole": "arn:aws:iam::123456789101:role/finance_users"
+    },
+    {
+      "iamRole": "arn:aws:iam::123456789101:role/default"
+    }
+  ]
+}
+```
+
+:::{list-table}
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:234]: Maybe add title for table

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 20]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+:header-rows: 1
+
+* - Property name
+  - Description
+* - `s3.security-mapping.enabled`
+  - Enable the security mapping feature.
+* - `s3.security-mapping.config-file`
+  - The JSON configuration file containing security mappings.
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:242]: Path to the JSON 

relative to ? or absolute? or both

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 21]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+  - The REST-endpoint URI containing security mappings.
+* - `s3.security-mapping.json-pointer`
+  - A JSON pointer (RFC 6901) to mappings inside the JSON retrieved from the
+    config file or REST-endpoint. The default is the root of the document.
+* - `s3.security-mapping.iam-role-credential-name`
+  - The name of the *extra credential* used to provide the IAM role.
+* - `s3.security-mapping.kms-key-id-credential-name`
+  - The name of the *extra credential* used to provide the KMS-managed key ID.
+* - `s3.security-mapping.refresh-period`
+  - How often to refresh the security mapping configuration.
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:253]: Something like 

Time duration between between repeated refreshing of the security mapping configuration

Comment (mosabua) [reply docs/src/main/sphinx/object-storage/file-system-s3.md:253]: could also link to the prop duration description.

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 22]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+* - `s3.security-mapping.colon-replacement`
+  - The character or characters to be used in place of the colon (`:`)
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:255]:
```suggestion
  - The character or characters to use instead of a colon (`:`)
```

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 23]
```diff
@@ -123,3 +123,138 @@ and secret keys, STS, or an IAM role:
...
+    character when specifying an IAM role name as an extra credential.
+    Any instances of this replacement value in the extra credential value
+    will be converted to a colon.
```
Comment (mosabua) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:258]:
```suggestion
    are converted to a colon.
```

**Review by mosabua — COMMENTED** [review]
Just a couple of doc suggestions

File: docs/src/main/sphinx/object-storage/file-system-s3.md (hunk 4 above)
Comment (electrum) [reply docs/src/main/sphinx/object-storage/file-system-s3.md:130]: I'll say "S3 locations" since it is bucket plus path within the bucket

File: docs/src/main/sphinx/object-storage/file-system-s3.md (hunk 6 above)
Comment (electrum) [reply docs/src/main/sphinx/object-storage/file-system-s3.md:137]: Do we go into this level of detail in the other documentation? Note that this documentation is copied over from the original, so we could do a more major cleanup as a separate task.

File: docs/src/main/sphinx/object-storage/file-system-s3.md (hunk 20 above)
Comment (electrum) [reply docs/src/main/sphinx/object-storage/file-system-s3.md:242]: It can be relative or absolute. I don't think we discuss "relative to what" in every place that uses additional files, but it's relative to the server configuration directory. Do we have a place that discusses that to link to?

**Comment by electrum** [comment]
@mosabua Thanks for the great review of the documentation. I believe I've addressed all of your comments.

**Comment by electrum** [comment]
@nineinchnick The tests are based on the legacy S3 support. 

TestS3SecurityMapping` has extensive tests to verify that we produce the correct `S3SecurityMappingResult` record, which is used to configure the S3 client.

What is not tested, either here or in the legacy version, is that the client configuration is correct. We would like to verify that the client is making the correct requests, i.e. using the correct access key, IAM role, KMS key ID, etc. This is what I investigated:

* Tracing: Doesn't work because AWS SDK tracing support for STS doesn't include the access key, role, etc.
* LocalStack: Create users and roles, then setup S3 policies such that different objects are only accessible to certain users or roles. Test via side effect that the correct values are sent -- the requests should fail if the wrong values are sent. I see that LocalStack supports IAM and STS, but I don't know if their security implementation is complete enough for this.
* Mocking: Use `MockWebServer` to validate requests and send mock responses for STS and S3. I have an initial version of this working, but didn't have time to complete it.

File: docs/src/main/sphinx/object-storage/file-system-s3.md (hunk 4 above)
Comment (mosabua) [reply docs/src/main/sphinx/object-storage/file-system-s3.md]: that works for me

File: docs/src/main/sphinx/object-storage/file-system-s3.md (hunk 6 above)
Comment (mosabua) [reply docs/src/main/sphinx/object-storage/file-system-s3.md:138]: We should .. just have not gotten around to fixing that up.

File: docs/src/main/sphinx/object-storage/file-system-s3.md (hunk 20 above)
Comment (mosabua) [reply docs/src/main/sphinx/object-storage/file-system-s3.md]: Not sure we have a generic place .. maybe just go with `Absolute or relative path to the JSON configuration file.`

**Comment by electrum** [comment]
@anusudarsan yes, I don’t plan to work on integration tests

File: docs/src/main/sphinx/object-storage/file-system-s3.md (hunk 6 above)
Comment (wendigo) [reply docs/src/main/sphinx/object-storage/file-system-s3.md:138]: Let's follow-up with the docs improvement after this PR is merged.

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemLoader.java [hunk 24]
```diff
@@ -0,0 +1,232 @@
...
+import java.util.function.Function;
+
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.trino.filesystem.s3.S3FileSystemConfig.RetryMode.getRetryStrategy;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+
+final class S3FileSystemLoader
+        implements Function<Location, TrinoFileSystemFactory>
```
Comment (wendigo) [inline lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemLoader.java:50]: Nit: this could be a dedicated interface

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SecurityMappingFileSystemFactory.java [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+    private final Map<Optional<S3SecurityMappingResult>, S3Client> clients = new ConcurrentHashMap<>();
+
+    public S3SecurityMappingFileSystemFactory(
+            S3SecurityMappingProvider mappingProvider,
+            S3ClientFactory clientFactory,
+            S3Context context,
+            Location location,
+            Executor uploadExecutor)
+    {
+        this.mappingProvider = mappingProvider;
```
Comment (wendigo) [inline lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SecurityMappingFileSystemFactory.java:47]: nit: rnn mappingProvider

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SecurityMappingProvider.java [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+
+    private Optional<String> getKmsKeyIdFromExtraCredential(ConnectorIdentity identity)
+    {
+        return kmsKeyIdCredentialName.map(name -> identity.getExtraCredentials().get(name));
+    }
+
+    private static Supplier<S3SecurityMappings> mappingsProvider(Supplier<S3SecurityMappings> supplier, Optional<Duration> refreshPeriod)
+    {
+        return refreshPeriod
+                .map(refresh -> Suppliers.memoizeWithExpiration(supplier::get, refresh.toMillis(), MILLISECONDS))
```
Comment (wendigo) [inline lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SecurityMappingProvider.java:136]: static import: memoize*

**Review by wendigo — APPROVED** [review]

**Comment by wendigo** [comment]
@electrum let's merge it and follow-up with code and test improvements

File: docs/src/main/sphinx/object-storage/file-system-s3.md [hunk 27]
```diff
@@ -123,3 +123,140 @@ and secret keys, STS, or an IAM role:
 * - `s3.external-id`
   - External ID for the IAM role trust policy when connecting to S3.
 :::
+
+## Security mapping
```
Comment (dain) [inline docs/src/main/sphinx/object-storage/file-system-s3.md:127]: Maybe move this "copy from legacy docs" stuff to a separate commit

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemLoader.java (hunk 24 above)
Comment (electrum) [reply lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemLoader.java:50]: We can introduce one later if it's useful

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SecurityMappingFileSystemFactory.java (hunk 25 above)
Comment (electrum) [reply lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3SecurityMappingFileSystemFactory.java:47]: Thanks, I missed that
